drivers: clock_control: Add clock_fixed driver

mfischer opened the pull request:
Add fixed-clock clock control driver. The new fixed-clock driver
allows for the possibility to share a driver between setups that
use fixed-clock and dynamic clock providers.

Since the change is hidden behind a Kconfig enable this is opt-in
for now.

Signed-off-by: Moritz Fischer <moritzf@google.com>

mfischer committed:
drivers: clock_control: Add clock_fixed_rate driver

Add fixed-clock clock control driver. This is a first step towards
making fixed-clocks a first-class citizen in the clock control
framework.

Since the change is hidden behind a Kconfig enable this is opt-in
for now.

Signed-off-by: Moritz Fischer <moritzf@google.com>

henrikbrixandersen commented on drivers/clock_control/Kconfig.fixed:
```diff
@@ -0,0 +1,10 @@
+
+config CLOCK_CONTROL_FIXED_CLOCK
+	bool "Fixed Clock Clock Control"
+	default n
```
I think the default should come from whether any fixed clock instances are enabled in the devicetree (see https://github.com/zephyrproject-rtos/zephyr/blob/8cb69f2c6b2e49047b46b79a446b6f65e2f9f9d5/drivers/clock_control/Kconfig.stm32#L39).

gmarull replied:
I agree, this needs to be opt-in. Many platforms won't use this at all, since it introduces overhead for little/no gain with the current clock control "API". However, it should at least depend on `DT_HAS_FIXED_CLOCK_ENABLED`.

mfischer replied:
Where is the DT_HAS_FIXED_CLOCK_ENABLED defined or what defines it?

henrikbrixandersen replied:
It is generated from parsing devicetree.

mfischer replied:
Done.

henrikbrixandersen commented on drivers/clock_control/clock_fixed.c:
```diff
@@ -0,0 +1,70 @@
+
+#define DT_DRV_COMPAT fixed_clock
+
+static int fixed_clk_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
```
This needs to turn on the input clock for the fixed clock, if any (as supported by the binding, see https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/clock/fixed-clock.yaml).

mfischer replied:
Oh! TIL. I didn't realize fixed clocks could have input clocks. Will fix.

mfischer replied:
Done. PTAL

mfischer replied:
@henrikbrixandersen do we have any use-case for this? I've implemented it but it seems kinda hard to wrap my mind around the idea of having a fixed-clock that depends on some other clock.

henrikbrixandersen replied:
I can't see why a fixed clock could _not_ have an input clock?

mfischer replied:
@henrikbrixandersen I think we should look at fixed-clock and fixed-factor clock separately.

For a fixed-factor clock, I completely agree, it should support parent clocks (since it is only a derived version of it).

For a fixed-clock (previous to this discussion) my admittedly Linux-based understanding was that it's an _always-on_ clock with a _fixed rate/frequency_.
Right now this is reflected by it being able to support them via compile time DT macros.

As such I'm struggling with use-cases for a parent clock in a _fixed-clock_ and propose to simplify the fixed-clock driver to be always-on with fixed rate.

henrikbrixandersen replied:
> As such I'm struggling with use-cases for a parent clock in a _fixed-clock_ and propose to simplify the fixed-clock driver to be always-on with fixed rate.

Okay, I see your point. Let's simplify the fixed-clock for now and leave out support for input clocks there. You'll need to remove the `clocks` property from the devicetree binding, then.

mfischer replied:
Done. PTAL.

henrikbrixandersen commented:
To complete this, we'd need a driver for `compatible = "fixed-factor-clock"` devices as well (https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/clock/fixed-factor-clock.yaml). Once both are in, we can start converting existing clock consumers to not special-case fixed clocks (example: https://github.com/zephyrproject-rtos/zephyr/blob/8cb69f2c6b2e49047b46b79a446b6f65e2f9f9d5/drivers/watchdog/wdt_mcux_wdog32.c#L189).

mfischer commented:
> To complete this, we'd need a driver for `compatible = "fixed-factor-clock"` devices as well (https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/clock/fixed-factor-clock.yaml). Once both are in, we can start converting existing clock consumers to not special-case fixed clocks (example:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/8cb69f2c6b2e49047b46b79a446b6f65e2f9f9d5/drivers/watchdog/wdt_mcux_wdog32.c#L189
> 
> ).

Ack. Will look at that next.

mfischer committed:
dts: bindings: clock: Fix fixed-clock binding

Remove `clocks` property for fixed-clock binding.
A fixed-clock should not have an input clock, since by
definition it's an always on fixed-rate clock.

Signed-off-by: Moritz Fischer <moritzf@google.com>

mfischer committed:
drivers: clock_control: Add clock_fixed_factor driver

Add fixed-factor-clock clock control driver. The new fixed-factor-clock
driver allows for defining clocks that output a fixed ratio of their
input rate.

Signed-off-by: Moritz Fischer <moritzf@google.com>

nordic-krch commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,96 @@
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sys);
+
+	return 0;
```
shouldn't it return `-ENOTSUP`?

mfischer replied:
Done. PTAL.

nordic-krch reviewed (commented):
It would be good to somehow ensure that it is being somehow compiled by CI. What about adding a configuration to the clock control test. Adding `-ENOTSUP` to `clock_control_off` may help to handle the test.

erwango commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,96 @@
+
+	ARG_UNUSED(sys);
+
+	*rate = config->rate;
```
I would expect that if this clock has no factor and fully depends from a known clock, then its rate also depend from the input clock.
Am I missing something ?

henrikbrixandersen replied:
That would be a `compatible = "fixed-factor-clock"` device. Perhaps both types of fixed clocks could be handled by the same driver, though?

mfischer replied:
@henrikbrixandersen do you think that's preferable to two separate drivers?

henrikbrixandersen replied:
@mfischer Depends on the amount of shared code between the two drivers. If most of the code is shared, I think one common driver with support for both compatibles is preferable.

mfischer replied:
The follow up commit to this one has the code for 'fixed-factor-clock'. I'll add tests and see if merging both into one driver buys me much.

henrikbrixandersen commented:
@mfischer Do you plan on picking this up again?

mfischer commented:
> @mfischer Do you plan on picking this up again?

Yes. Work got busy last couple of weeks. It's on my list.

erwango commented on drivers/clock_control/clock_fixed_factor.c:
```diff
@@ -0,0 +1,104 @@
+		NULL, \
+		NULL, \
+		&fixed_factor_clock_config_##idx, \
+		PRE_KERNEL_1, \
+		CONFIG_CLOCK_CONTROL_INIT_PRIORITY, \
```
About the priority:
Since client of `clock_control` API, it should be done after init of clock_control driver.
But yet, at least on STM32, we need fixed clocks to be initialized in the early phase since some other clocks will depend on fixed clocks availability (PLL for instance).
So we'd need similar driver enabled for PLL as well, that would be initialized after fixed (factor) clocks.
Then a driver for sysclock which usually requires PLL.
And finally you'll need to configure peripheral busses.
(Note that this is a very simple approach as you usually also need to perform other actions like update Flash latency before and/or after this whole sequence)

Taking this into account, I wonder about the whole plan for getting this driver to be really useful.
Or we should assume that it won't ever be compatible with some platforms.

@mfischer can you comment on that part?

Note: I'm setting this comment as "blocking" to trigger the discussion, but it is not meant to be blocking in the end or even require changes (depending on the output of the discussion).

mfischer replied:
@erwango I've updated change set to simplify the fixed-rate clock driver, I don't think it should have any external dependencies left. What would be the earliest priority you'd want the driver to initialize?

Note: For a fixed-rate clock you could in theory always bypass the clock control API and pull the rate / frequency out of DT, since, by definition it's always on and has a fixed rate.

The use-case I'm trying to enable is to allow for drivers to runtime derive their input clock frequency - independently if they're being fed by a fixed-rate clock or another clock control API provider.

erwango replied:
> @erwango I've updated change set to simplify the fixed-rate clock driver, I don't think it should have any external dependencies left. What would be the earliest priority you'd want the driver to initialize?

Assuming: 
- You don't need clock_control API at driver init
- Fixed clocks are needed for core clock configuration (as part of the core clock tree)
This driver needs to be initialized before the piece of code configuring core clock (`clock_control` in STM32 case).

> The use-case I'm trying to enable is to allow for drivers to runtime derive their input clock frequency - independently if they're being fed by a fixed-rate clock or another clock control API provider.

This is already possible with existing API. What are the limitation you're seeing ?

henrikbrixandersen replied:
> This is already possible with existing API. What are the limitation you're seeing ?

You cannot call `clock_control_get_rate()` on a fixed (factor) clock in the tree as it is today, thus requiring all drivers with a `clocks` devicetree property having to check for a fixed (factor) clock at compile-time and use the clock rate directly - or defer to run-time and call `clock_control_get_rate()`.

erwango replied:
Ok, right. I'm still assuming that we only have one `clock_control` driver centralizing all clock related function (which makes this possible), but I understand this will tend to be old fashioned in medium term.

mfischer replied:
@erwango if you want to keep the same binding for fixed clocks I'm not sure how you'd do that with a single driver. Can you elaborate?

erwango replied:
Well this is how it it handled today in STM32 clock_control driver for instance. There are multiple clocks compatibles, but only one driver handles the whole information to build the clock tree.

I understand that this is a shift of paradigm and for this driver to be usable in most clocks tree it requires a complete clock framework around it.
Let me synthesize the assumptions around this driver:
- Clock initialization should be done prior to the use of this driver
- It cannot be called in the core clock tree initialization phase if such phase is done as part of a 'main' clock_control driver init.
Is that correct ?

erwango reviewed (changes requested):
Questions on driver usability/limitations and global plan.

mfischer commented on tests/drivers/clock_control/fixed_clock/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_ZTEST=y
```
Done.

mfischer commented on tests/drivers/clock_control/fixed_clock/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  drivers.clock.clock_control_fixed_clock:
+    tags: drivers
+    platform_allow: native_posix
```
What would you suggest? I'm pretty new to twister / ZTEST, so open to feedback.

erwango replied:
Is there any hw compatible with this today ?
If the point is to run a sample with test-clocks, native_posix is good enough, IMO.

mfischer replied:
Can we leave this as is for now? It's purely a software concept we're testing for now.

mfischer commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,96 @@
+static enum clock_control_status fixed_rate_clk_get_status(const struct device *dev,
+							   clock_control_subsys_t sys)
+{
+	return CLOCK_CONTROL_STATUS_ON;
```
With removed input clocks no longer applies.

mfischer commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,96 @@
+	if (!config->clk_dev)
+		return 0;
+
+	return clock_control_on(config->clk_dev, config->clk_subsys);
```
No longer applies.

mfischer commented on drivers/clock_control/clock_fixed_factor.c:
```diff
@@ -0,0 +1,104 @@
+{
+	const struct fixed_factor_clock_config *config = dev->config;
+
+	return clock_control_on(config->clk_dev, config->clk_subsys);
```
Done.

gmarull commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,79 @@
+	return 0;
+}
+
+#define FIXED_CLK_INIT(idx) \
+	static const struct fixed_rate_clock_config fixed_rate_clock_config_##idx = { \
+		.rate = DT_INST_PROP(idx, clock_frequency), \
+	}; \
+	DEVICE_DT_INST_DEFINE(idx, \
+		fixed_rate_clk_init, \
+		NULL, \
+		NULL, \
+		&fixed_rate_clock_config_##idx, \
+		PRE_KERNEL_1, \
+		CONFIG_CLOCK_CONTROL_INIT_PRIORITY, \
+		&fixed_rate_clk_api \
+	);
+DT_INST_FOREACH_STATUS_OKAY(FIXED_CLK_INIT)
```
please align `\`, or try with clang-format.

mfischer replied:
Done. PTAL

gmarull commented on tests/drivers/clock_control/fixed_clock/app.overlay:
```diff
@@ -0,0 +1,17 @@
+		test_fixed_rate_clk0: clk {
+			compatible = "fixed-clock";
+			clock-frequency = <100>;
+			#clock-cells = <0>;
```
clock-cells needs to be removed

mfischer replied:
That's of course a mistake. Still needs #clock-cells = <0>, added back to binding. Good catch. Thanks!.

gmarull commented on drivers/clock_control/clock_fixed_factor.c:
```diff
@@ -0,0 +1,104 @@
+	.get_rate = fixed_factor_clk_get_rate
+};
+
+#define INPUT_SUBSYS_OR_ALL(idx) \
+	COND_CODE_1(DT_INST_NUM_CLOCKS(idx), \
+		(DT_INST_PHA_BY_IDX_OR(idx, clocks, 0, name, \
+				      CLOCK_CONTROL_SUBSYS_ALL)), \
+		    (CLOCK_CONTROL_SUBSYS_ALL))
```
can't understand what's going on here, can you please explain?

mfischer replied:
If you have a input clock and if that input clock has a phandle, and if that phandle does have a 'name' cell grab the 'first' (starting at 0) cell which is the subsystem <&foo_clk FOO_CLK_NAME>, otherwise use CLOCK_CONTROL_SUBSYS_ALL (e.g. for a parent clock with #cells = <0>, i.e. a fixed_rate_clock).

I just couldn't find a better way of dealing with #clock-cells=<0> and #clock-cells=<1>. Obviously this fails all other cases with #clock-cells > 1.

erwango commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,79 @@
+static int fixed_rate_clk_on(const struct device *dev,
+			     clock_control_subsys_t sys)
+{
+	ARG_UNUSED(dev);
```
If `fixed_rate_clk_init` is empty and you aren't doing anything on `fixed_rate_clk_on`, where is the clock actually enabled on the SoC ?

mfischer replied:
It's an always-on clock (e.g. external), it doesn't get enabled, or it might have gotten enabled by a previous boot loader.

mfischer commented on drivers/clock_control/clock_fixed_rate.c:
```diff
@@ -0,0 +1,79 @@
+	return -ENOTSUP;
+}
+
+static enum clock_control_status fixed_rate_clk_get_status(const struct device *dev,
```
My goal long term is that as a consumer of the API I don't have to think about what type of clock I look at. If you have another idea on how to do that, I'd be happy to make the change.

mfischer commented on tests/drivers/clock_control/fixed_clock/app.overlay:
```diff
@@ -0,0 +1,24 @@
+			compatible = "fixed-factor-clock";
+			#clock-cells = <0>;
+			clock-mult = <0x1>;
+			clock-div = <0x1>;
```
Done. PTAL

henrikbrixandersen commented:
Where are we with this PR? I'd really like to see those fixed clock drivers make it into the tree.

mfischer commented:
> @mfischer Do you plan on picking this up again?

I'm currently quite busy with other stuff. I can get back to it eventually, but right now don't have the bandwidth to work on this.
